chain/network/src/peer/peer_actor.rs

```diff
@@ -1013,6 +1009,58 @@ impl StreamHandler<Result<Vec<u8>, ReasonForBan>> for PeerActor {
                         Some(self.throttle_controller.clone()),
                     ));
             }
+            (PeerStatus::Ready, PeerMessage::SyncAccountsData(msg)) => {
+                let pms = self.peer_manager_state.clone();
+                // In case a full sync is requested, immediately send what we got.
+                // It is a microoptimization: we do not send back the data we just received.
+                if msg.requesting_full_sync {
+                    self.send_message_or_log(&PeerMessage::SyncAccountsData(SyncAccountsData {
+                        requesting_full_sync: false,
+                        incremental: false,
+                        accounts_data: pms.accounts_data.dump(),
+                    }));
+                }
+                async move {
+                    // Early exit, if there is no data in the message.
+                    if msg.accounts_data.is_empty() {
+                        return None;
+                    }
+                    // Check if the chain progressed to the next epoch.
+                    let chain_info = pms.clone().get_chain_info().await;
+                    pms.accounts_data
+                        .set_epochs(vec![&chain_info.this_epoch, &chain_info.next_epoch]);
+                    // Verify and add the new data to the internal state.
+                    let (new_data, err) = pms.accounts_data.clone().insert(msg.accounts_data).await;
+                    // Broadcast any new data we have found.
+                    // TODO(gprusak): this should be rate limited - diffs should be aggregated,
+                    // unless there was no recent broadcast of this type.
+                    if new_data.len() > 0 {
+                        pms.broadcast_message(SendMessage {
+                            message: PeerMessage::SyncAccountsData(SyncAccountsData {
+                                incremental: true,
+                                requesting_full_sync: false,
+                                accounts_data: new_data,
+                            }),
+                            context: Span::current().context(),
+                        })
+                        .await;
+                    }
+                    err.map(|err| match err {
```

Comment:
maybe it's worth it to implement a function in accounts_data for this conversion?

Reply:
accounts_data by design is not aware of peer-banning logic, it just reports problems. I wanted to keep banning logic in PeerActor & PeerManagerActor.